shared/types/time.go

```diff
@@ -52,3 +59,181 @@ func (interval RepeatingTimeInterval) Includes(testTime Moment) bool {
 	}
 	return false
 }
+
+type Interval struct {
+	start time.Time
+	end   time.Time
+}
+
+func (interval Interval) WithDate(year int, month time.Month, day int) Interval {
+	return Interval{
+		start: time.Date(
+			year, month, day,
+			interval.start.Hour(), interval.start.Minute(),
+			0, 0, interval.start.Location()),
+		end: time.Date(
+			year, month, day,
+			interval.end.Hour(), interval.end.Minute(),
+			0, 0, interval.end.Location()),
+	}
+}
+
+type Intervals []Interval
+
+func (intervals Intervals) Len() int {
+	return len(intervals)
+}
+func (intervals Intervals) Swap(i, j int) {
+	intervals[i], intervals[j] = intervals[j], intervals[i]
+}
+func (intervals Intervals) Less(i, j int) bool {
+	return intervals[i].start.Sub(intervals[j].start) < 0
+}
+
+func (repeatingTimeIntervals RepeatingTimeIntervals) weekdayIntervals() map[time.Weekday][]Interval {
```

Comment:
I'll just assume that you have enough tests and that there are no bugs in this.  :-P